### PR TITLE
Cortx 30426 Implement mix mode io operation using s3bench.

### PR DESCRIPTION
**Pull request by DhobaleVishal**
## Problem Statement

-   Problem statement

## Design

-   For Bug, Describe the fix here.
-   For new test addition post the link for design

## Coding

   Checklist for Author

-   [ ]  Coding conventions are followed and code is formatted (e.g. using pycharm in-built formatter)

## Testing

  Checklist for Author

-   [ ]  New/Affected tests are executed
-   [ ]  Attach test execution logs

## Review 

  Checklist for Author

-   [x]  JIRA number/GitHub Issue added to PR
-   [x]  PR is self reviewed
-   [x]  Jira and state/status is updated and JIRA is updated with PR link
-   [x]  Check if the description is clear and explained

## Documentation

  Checklist for Author

-   [ ]  Changes done to ReadMe

### CODE REVIEW

File: workload/s3/mix_io/type4_mix_object_crud_operations.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+test_1:
```
Comment (VikasKumarC) [inline workload/s3/mix_io/type4_mix_object_crud_operations.yaml:1]: May we have  comments to explain the parameters purpose

File: scripts/s3/mixs3io/mix_object_operations.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        if not S3bench.install_s3bench():
+            raise Exception("s3bench tool is not installed.")
+        if kwargs.get("duration"):
+            self.finish_time = datetime.now() + kwargs.get("duration")
+        else:
+            self.finish_time = datetime.now() + timedelta(hours=int(100 * 24))
+        self.define_variables(**kwargs)
+
+    @classmethod
+    def define_variables(cls, **kwargs):
```
Comment (VikasKumarC) [inline scripts/s3/mixs3io/mix_object_operations.py]: Update/Explain parameters in docstring

File: scripts/s3/mixs3io/mix_object_operations.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+                self.log.exception("bucket url: {%s}\nException: {%s}", self.s3_url, err)
+                assert False, f"bucket url: {self.s3_url}\nException: {err}"
+            if (self.finish_time - datetime.now()).total_seconds() < MIN_DURATION:
+                self.delete_s3_bucket(self.bucket_name, force=True)
+                return True, "Bucket operation execution completed successfully."
+            self.log.info("iteration %s is completed...", iteration)
+            iteration += 1
+
+    def get_sample_details(self, file_size):
+        """Get all stats to be used in IO."""
```
Comment (VikasKumarC) [inline scripts/s3/mixs3io/mix_object_operations.py]: update doc string , explain function params

File: scripts/s3/mixs3io/mix_object_operations.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        assert number_sample > self.sessions, "Number of samples should be greater then session."
+        read_sample = int(int(self.total_storage / 100 * self.read_percentage) / file_size) if \
+            self.read_percentage else number_sample
+        assert read_sample > self.sessions, "Number of samples should be greater then session."
+        delete_sample = int(int(self.total_storage / 100 * self.delete_percentage) / file_size)  \
+            if self.delete_percentage else number_sample
+        assert delete_sample > self.sessions, "Number of samples should be greater then session."
+        return storage_to_fill, number_sample, read_sample, delete_sample
+
+    def s3bench_cmd(self, object_size, number_sample):
```
Comment (VikasKumarC) [inline scripts/s3/mixs3io/mix_object_operations.py]: update doc string with function params

File: scripts/s3/mixs3io/mix_object_operations.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    def execute_validate_run(self, cmd):
+        """Execute command and validate the execution report and logs."""
+        cmd += self.cmd_reporting_params()
+        status, resp = run_local_cmd(cmd)
+        assert status, f"Failed execute '{cmd}', response: {resp}"
+        status, resp = S3bench.check_log_file_error(self.report_path, self.log_path)
+        assert status, f"Observed failures for '{cmd}', response: {resp}"
+
+    def write_data(self, object_size, number_sample, validate=False):
+        """Write data to s3."""
```
Comment (VikasKumarC) [inline scripts/s3/mixs3io/mix_object_operations.py]: update docstring

File: scripts/s3/mixs3io/mix_object_operations.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        cmd = self.s3bench_cmd(object_size, number_sample)
+        if validate:
+            cmd += " -skipRead -skipCleanup -validate"
+        else:
+            cmd += " -skipRead -skipCleanup"
+        self.execute_validate_run(cmd)
+        self.total_written_data += object_size + number_sample
+        self.log.info("Upload completed...")
+
+    def read_data(self, object_size, number_sample, validate=False):
```
Comment (VikasKumarC) [inline scripts/s3/mixs3io/mix_object_operations.py]: update doc string to explain function params

File: scripts/s3/mixs3io/mix_object_operations.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+from time import perf_counter_ns
+
+from src.commons.constants import LATEST_LOG_PATH
+from src.commons.constants import MIN_DURATION
+from src.commons.utils.corio_utils import run_local_cmd
+from src.libs.s3api.s3_bucket_ops import S3Bucket
+from src.libs.s3api.s3_object_ops import S3Object
+from src.libs.tools.s3bench import S3bench
+
+
```
Comment (VikasKumarC) [inline scripts/s3/mixs3io/mix_object_operations.py:34]: Update doc string for methods in class, explain params

**Review by VikasKumarC — COMMENTED** [review]
@DhobaleVishal  Please check and update.

File: workload/s3/mix_io/type4_mix_object_crud_operations.yaml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    TEST_ID: TEST-40042
+    object_size:
+        start: 0bytes
+        end: 1Gib
+    write_percentage: 30  # Percentage of data to fill the storage.
+    read_percentage: 30  # Percentage of data to be used to read from storage.
+    delete_percentage: 15  # Percentage of data to be deleted from storage.
+    cleanup_percentage: 85  # Percentage of data once reached then cleanup all data.
+    total_storage_size: 1TiB  # Total available storage from cluster or user to be used.
```
Comment (priyankaborawake) [inline workload/s3/mix_io/type4_mix_object_crud_operations.yaml]: This should be calculated from total capacity available in hctl status.

File: workload/s3/mix_io/type4_mix_object_crud_operations.yaml [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    min_runtime: 30d  # Minimum execution durations to mark workload pass.
+    sessions_per_node: 20  # Number of sessions per node.
+    background_delete: False  # Enable/Disable background delete(Cortx specific).
```
Comment (priyankaborawake) [inline workload/s3/mix_io/type4_mix_object_crud_operations.yaml:14]: What will this parameter be used for ?

Comment (DhobaleVishal) [reply workload/s3/mix_io/type4_mix_object_crud_operations.yaml:14]: This is used to enable/disable backgroud delete as per type-4/5 scenario. We will add support once available in cortx.

File: workload/s3/mix_io/type4_mix_object_crud_operations.yaml (hunk 8 above)
Comment (DhobaleVishal) [reply workload/s3/mix_io/type4_mix_object_crud_operations.yaml]: User capacity might be indefinite hence we have added this option for user input.

**Comment by priyankaborawake** [comment]
@DhobaleVishal Was this finalized to use S3bench in CORIO. Is there any documented communication on this.

File: workload/s3/mix_io/type4_mix_object_crud_operations.yaml (hunk 8 above)
Comment (DhobaleVishal) [reply workload/s3/mix_io/type4_mix_object_crud_operations.yaml]: Added code to fetch it from cluster.

**Comment by DhobaleVishal** [comment]
@VikasKumarC, @priyankaborawake Addressed & incorporated review comments, pls have a look.

**Comment by VikasKumarC** [comment]
> @DhobaleVishal Was this finalized to use S3bench in CORIO. Is there any documented communication on this.
@priyankaborawake  Yes this was finalized and its a consensus to everyone to use not only s3bench but any tool as a wrapper.

**Review by VikasKumarC — APPROVED** [review]
Comments addressed and Tested with execution.
Fine to merge with dev.

**Pull request merged**